### PR TITLE
Assistant for Unified Package

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Part of #15122

Updates the assistant with logic to actually get a unified package. Also shifts from fire to jsonargparse for CLI execution as fire had issues with argument parsing.

This only ports parts of the changed assistant logic from #15122 as the remaining parts will be ported within the PRs related to these specific fixes!

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca @akihironitta @borda

### CODE REVIEW

File: .actions/assistant.py [hunk 1]
```diff
@@ -168,6 +185,52 @@ def mirror_pkg2source(pypi_folder: str, src_folder: str) -> None:
...
+                fp_new = fp
+
+            print(fp_new, fp)
+            with open(fp_new, "w", encoding="utf-8") as fo:
+                fo.writelines(py)
+
 
 if __name__ == "__main__":
-    fire.Fire(AssistantCLI)
+    jsonargparse.CLI(AssistantCLI, as_positional=True)
```
Comment (Borda) [inline .actions/assistant.py:236]: What was the problem with Fire?

Comment (otaj) [reply .actions/assistant.py:236]: different parsing of strings if string contained `.` or not

**Review by otaj — APPROVED** [review]

File: .actions/assistant.py [hunk 2]
```diff
@@ -60,6 +60,23 @@ def _load_py_module(name: str, location: str) -> ModuleType:
     return py
 
 
+def _retrieve_files(directory: str, *ext: str):
+    subfolders, files = [directory], []
+
+    ext = tuple(e.lower() for e in ext)
```
Comment (Borda) [inline .actions/assistant.py:66]: Do we have capital extensions anywhere?

File: .actions/assistant.py [hunk 3]
```diff
@@ -168,6 +185,52 @@ def mirror_pkg2source(pypi_folder: str, src_folder: str) -> None:
...
+        )
+
+        if target_dir is None:
+            target_dir = source_dir
+
+        ls = _retrieve_files(source_dir)
+
+        for fp in ls:
+            if not os.path.isfile(fp):
+                continue
```
Comment (Borda) [inline .actions/assistant.py:208]:
```suggestion
```
this seems redundant given your `_retrieve_files`

File: .actions/assistant.py [hunk 4]
```diff
@@ -168,6 +185,52 @@ def mirror_pkg2source(pypi_folder: str, src_folder: str) -> None:
...
+                    if os.path.abspath(fp) != os.path.abspath(fp_new):
+                        shutil.copy2(fp, fp_new)
+                continue
+
+            with open(fp, encoding="utf-8") as fo:
+                py = fo.readlines()
+
+            for _source_import, _target_import in zip(source_imports, target_imports):
+                for i, ln in enumerate(py):
+                    py[i] = ln.replace(_source_import, _target_import)
```
Comment (Borda) [inline .actions/assistant.py:222]: This may break replacements in function names.

Comment (otaj) [reply .actions/assistant.py:222]: That's why https://github.com/Lightning-AI/lightning/pull/15208 has to go in first

Comment (Borda) [reply .actions/assistant.py:222]:
```suggestion
                    py[i] = re.sub(rf"(?!_){_source_import}(?!_)", _target_import, ln)
```

File: .actions/assistant.py (hunk 3 above)
Comment (justusschock) [reply .actions/assistant.py:208]: True, this was just a security measure before `_retrieve_files` was reimplemented!

File: .actions/assistant.py (hunk 2 above)
Comment (justusschock) [reply .actions/assistant.py:66]: because globbing with a wildcard only is error prone and slower.

File: .actions/assistant.py (hunk 4 above)
Comment (otaj) [reply .actions/assistant.py:220]: See https://github.com/Lightning-AI/lightning/pull/15105#discussion_r1000355354, I really don't think your regex would guard against instances of it properly

Comment (justusschock) [reply .actions/assistant.py:220]: @Borda can you push this to #15122 to see if this works in CI?

Comment (otaj) [reply .actions/assistant.py:220]: It will work in #15122, because we already changed all the occurrences which would be damaged

Comment (Borda) [reply .actions/assistant.py:220]: yes there are a few which need to be edited, https://github.com/Lightning-AI/lightning/commit/0a17aba13e6087dbd0cc5cf932236491e7d38763

File: .actions/requirements.txt [hunk 5]
```diff
@@ -1,3 +1,3 @@
-fire
+jsonargparse
```
Comment (Borda) [inline .actions/requirements.txt:1]: any version?

Comment (justusschock) [reply .actions/requirements.txt:1]: doesn't really matter here.

Comment (Borda) [reply .actions/requirements.txt:1]:
```suggestion
jsonargparse >= 4.15.0
```

File: .actions/assistant.py (hunk 2 above)
Comment (Borda) [reply .actions/assistant.py:66]: I do not think we are here playing with ms :)
We can change it later on...

File: .actions/assistant.py [hunk 6]
```diff
@@ -168,6 +185,50 @@ def mirror_pkg2source(pypi_folder: str, src_folder: str) -> None:
...
+                for i, ln in enumerate(py):
+                    py[i] = re.sub(rf"(?!_){_source_import}(?!_)", _target_import, ln)
+
+            if target_dir:
+                fp_new = fp.replace(source_dir, target_dir)
+                os.makedirs(os.path.dirname(fp_new), exist_ok=True)
+            else:
+                fp_new = fp
+
+            print(fp_new, fp)
```
Comment (Borda) [inline .actions/assistant.py:228]:
```suggestion
```

**Review by Borda — APPROVED** [review]

File: .actions/assistant.py [hunk 7]
```diff
@@ -60,6 +60,23 @@ def _load_py_module(name: str, location: str) -> ModuleType:
...
+
+    ext = tuple(e.lower() for e in ext)
+
+    while subfolders:
+        for p in os.scandir(directory):
+            if os.path.isdir(p):
+                subfolders.append(p)
+
+            if os.path.isfile(p):
+                if not ext or os.path.split(p.name)[1].lower() in ext:
```
Comment (carmocca) [inline .actions/assistant.py:74]: this should avoid possible problems with `.pyc` files matching `.py`

```suggestion
                if not ext or any(os.path.split(p.name)[1].lower().endswith(e) for e in ext):
```

File: .actions/assistant.py [hunk 8]
```diff
@@ -168,6 +185,49 @@ def mirror_pkg2source(pypi_folder: str, src_folder: str) -> None:
...
+
+        if target_dir is None:
+            target_dir = source_dir
+
+        ls = _retrieve_files(source_dir)
+
+        for fp in ls:
+            if not fp.endswith(".py"):
+                if not fp.endswith(".pyc"):
+                    fp_new = fp.replace(source_dir, target_dir)
```
Comment (carmocca) [inline .actions/assistant.py:209]: what is this block for? markdown files and such?

File: .actions/requirements.txt [hunk 9]
```diff
@@ -1,3 +1,3 @@
-fire
+jsonargparse >= 4.15.0
```
Comment (carmocca) [inline .actions/requirements.txt:1]: nit: usually we don't use spaces

```suggestion
jsonargparse>=4.15.0
```

File: .actions/assistant.py (hunk 8 above)
Comment (otaj) [reply .actions/assistant.py:218]: Exactly. Basically anything that isn't python. JS, MD, images, etc

Comment (carmocca) [reply .actions/assistant.py:218]: Okay. I would say this is simpler to read then


```python
            if fp.endswith(".py"):
                with open(fp, encoding="utf-8") as fo:
                    py = fo.readlines()

                for _source_import, _target_import in zip(source_imports, target_imports):
                    for i, ln in enumerate(py):
                        py[i] = re.sub(rf"(?!_){_source_import}(?!_)", _target_import, ln)

                if target_dir:
                    fp_new = fp.replace(source_dir, target_dir)
                    os.makedirs(os.path.dirname(fp_new), exist_ok=True)
                else:
                    fp_new = fp

                with open(fp_new, "w", encoding="utf-8") as fo:
                    fo.writelines(py)
            elif not fp.endswith("pyc") :
                fp_new = fp.replace(source_dir, target_dir)
                os.makedirs(os.path.dirname(fp_new), exist_ok=True)
                if os.path.abspath(fp) != os.path.abspath(fp_new):
                    shutil.copy2(fp, fp_new)
```

File: .actions/assistant.py [hunk 10]
```diff
@@ -168,6 +194,49 @@ def mirror_pkg2source(pypi_folder: str, src_folder: str) -> None:
...
+                    os.makedirs(os.path.dirname(fp_new), exist_ok=True)
+                    if os.path.abspath(fp) != os.path.abspath(fp_new):
+                        shutil.copy2(fp, fp_new)
+                continue
+
+            with open(fp, encoding="utf-8") as fo:
+                py = fo.readlines()
+
+            for _source_import, _target_import in zip(source_imports, target_imports):
```
Comment (carmocca) [inline .actions/assistant.py:227]: Why the preceding underscores?

**Review by carmocca — APPROVED** [review]

File: .actions/assistant.py (hunk 8 above)
Comment (Borda) [reply .actions/assistant.py:218]: pls send it as suggestion :)

File: .actions/assistant.py (hunk 10 above)
Comment (Borda) [reply .actions/assistant.py:227]: no need, can be edited :)

**Pull request merged**